plugins/microservices/src/json_compare.cpp

```diff
@@ -0,0 +1,189 @@
+
+            const json* json_object_2{};
+
+            if (std::strcmp(handle_1, handle_2) == 0) {
+                json_object_2 = &json_object_1;
+            }
+            else {
+                auto* object = is::find(handle_2);
+                if (!object) { // NOLINT(readability-implicit-bool-conversion)
+                    log_msi::error("Invalid argument: JSON handle [{}] is not associated with any object.", handle_2);
+                    return SYS_INVALID_INPUT_PARAM;
+                }
+
+                json_object_2 = &boost::any_cast<json&>(*object);
+            }
+
+            const json* json_value_2{};
+
+            try {
+                const auto* p = static_cast<const char*>(_json_pointer_2->inOutStruct);
+                // NOLINTNEXTLINE(readability-implicit-bool-conversion)
+                log_msi::debug("Second JSON pointer (address=[{}], value=[{}])", static_cast<const void*>(p), (p ? p : "<nullptr>"));
+
+                // TODO Should this result in an error if the the JSON pointer is invalid?
+                json_value_2 = &json_object_2->at(json::json_pointer{p});
+            }
+            catch (const json::out_of_range&) {
+                fillStrInMsParam(_cmp_result, "<irods::json::out_of_range>");
+                return 0;
+            }
+
+            char* result{};
+
+            // clang-format off
+            if      (*json_value_1 == *json_value_2) { result = strdup("<irods::json::equivalent>"); }
+            else if (*json_value_1  < *json_value_2) { result = strdup("<irods::json::less>"); }
+            else if (*json_value_1  > *json_value_2) { result = strdup("<irods::json::greater>"); }
+            else                                     { result = strdup("<irods::json::unordered>"); }
+            // clang-format on
+
+            // The NREP defines the list() construct as either a StrArray or IntArray.
+            // For this microservice, the type if StrArray because we're dealing with strings.
+            fillStrInMsParam(_cmp_result, result);
+        }
+        catch (const std::exception& e) {
+            log_msi::error("Caught exception while comparing JSON objects. [{}]", e.what());
+            return SYS_LIBRARY_ERROR;
+        }
+
+        return 0;
+    } // msi_impl
+
+    template <typename... Args, typename Function>
+    auto make_msi(const std::string& name, Function func) -> irods::ms_table_entry*
+    {
+        auto* msi = new irods::ms_table_entry{sizeof...(Args)}; // NOLINT(cppcoreguidelines-owning-memory)
+        msi->add_operation(name, std::function<int(Args..., ruleExecInfo_t*)>(func));
+        return msi;
+    } // make_msi
+} // anonymous namespace
+
+extern "C"
+auto plugin_factory() -> irods::ms_table_entry*
+{
+    return make_msi<MsParam*, MsParam*, MsParam*, MsParam*, MsParam*>("msi_json_compare", msi_impl);
+} // plugin_factory
+
+#ifdef IRODS_FOR_DOXYGEN
+/// \brief Compares two elements within one or two JSON structures.
+///
```

Comment:
Document that the parsed JSON structures obtained via `msi_json_parse_*` will not be available in delay rules.

Reply:
Done.